py/dml/codegen.py

```diff
@@ -394,6 +395,462 @@ def memoization_common_prelude(name, site, outp, throws, make_ref):
                      make_ref('ran', TInt(8, True)),
                      mkCompound(site, unrun + has_run + running))]
 
+class TypeSequenceInfo:
+    """Bookkeeping surrounding a realtyped unique type sequence"""
+    def __init__(self, types, uniq):
+        self.types = types
+        self.uniq = uniq
+        self.after_on_hooks = {}
+        self.struct = (TStruct({f'comp{i}': typ
+                                 for (i, typ) in enumerate(types)},
+                                label=f'_typeseq_{self.uniq}')
+                       if types else None)
+        type_keys = ', '.join(f"'{typ.key()}'" for typ in types)
+        self.string_key = f'({type_keys})'
+
+    def get_after_on_hook(self, aoh_prim_key, param_to_msg_comp, no_params,
+                          createnew=False):
+        param_to_msg_comp_key = tuple(param_to_msg_comp.get(i)
+                                      for i in range(no_params))
+        try:
+            return self.after_on_hooks[(aoh_prim_key, param_to_msg_comp_key)]
+        except KeyError:
+            if not createnew:
+                raise
+            info = after_on_hook_info_constructors[type(aoh_prim_key)](
+                self, aoh_prim_key, param_to_msg_comp)
+            self.after_on_hooks[(aoh_prim_key, param_to_msg_comp_key)] = info
+            return info
+
+class TypeSequenceInfoDict:
+    """Mappings from a unique type sequence to its TypeSequenceInfo"""
+    def __init__(self):
+        self._dict = {}
+        self._list = []
+
+    def get(self, index, realtyped=False, createnew=False):
+        index = tuple(index if realtyped
+                      else (safe_realtype(typ) for typ in index))
```

Comment:
I'd say a redundant realtype is cheap enough to be done unconditionally. Or if number of callers is small, just leave the class+realtype mapping to the caller.

If you insist not to then I'm fine with that, but in that case consider:
```
if not realtyped:
  index = ...
```

Reply:
We need the forced tuplification anyway. I can agree with redundant realtyping if you're fine with that.

Reply:
Now doing redundant realtyping.